cli/CommandExecutor.cpp

```diff
@@ -0,0 +1,200 @@
+/**
+ *   Copyright 2011-2015 Quickstep Technologies LLC.
+ *   Copyright 2015-2016 Pivotal Software, Inc.
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#include "cli/CommandExecutor.hpp"
+
+#include <cstddef>
+#include <cstdio>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "catalog/CatalogAttribute.hpp"
+#include "catalog/CatalogDatabase.hpp"
+#include "catalog/CatalogRelation.hpp"
+#include "catalog/CatalogRelationSchema.hpp"
+#include "cli/PrintToScreen.hpp"
+#include "parser/ParseStatement.hpp"
+#include "utility/PtrVector.hpp"
+#include "utility/Macros.hpp"
+#include "utility/SqlError.hpp"
+
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+using std::fprintf;
+using std::fputc;
+using std::fputs;
+using std::size_t;
+using std::vector;
+using std::string;
+
+namespace quickstep {
+
+void CommandExecutor::executeCommand(const ParseStatement &statement,
+                                     const CatalogDatabase *catalog_database,
+                                     FILE *out) {
+  const ParseCommand &command = static_cast<const ParseCommand &>(statement);
+  const PtrVector<ParseString> *arguments = command.arguments();
+  std::string command_str = command.command()->value();
+  if (command_str.compare("\\dt") == 0) {
+    executeDescribeDatabase(arguments, catalog_database, out);
+  } else if (command_str.compare("\\d") == 0) {
+    if (arguments->size() == 0) {
+      executeDescribeDatabase(arguments, catalog_database, out);
+    } else {
+      executeDescribeTable(arguments, catalog_database, out);
+    }
+  } else {
+    THROW_SQL_ERROR_AT(command.command()) << "Invalid Command";
+  }
+}
+
+void CommandExecutor::executeDescribeDatabase(
+    const PtrVector<ParseString> *arguments,
+    const CatalogDatabase *catalog_database, FILE *out) {
+  std::size_t relation_count = 0;
+  // Column width initialized to 6 to take into account the header name
+  // and the column value table
+  int max_column_width = 6;
```

Comment:
The alignment of the output is off, and we should make sure the output is nicely aligned along the column. 

e.g. Currently when we list the database table for SSB, it shows (notice the header row is mis-aligned): 

```quickstep> \d
       List of relations

 Name    | Type |
+---------+---------+
 lineorder| table    
 part     | table    
 supplier | table    
 customer | table    
 ddate    | table 
``` 
PostgreSQL is a good model to follow. Here are some output from it.

```
jignesh=# \d
          List of relations
 Schema |   Name    | Type  |  Owner  
--------+-----------+-------+---------
 public | customer  | table | jignesh
 public | ddate     | table | jignesh
 public | lineorder | table | jignesh
 public | part      | table | jignesh
 public | supplier  | table | jignesh
(5 rows)
```
and
```
jignesh=# \d
                        List of relations
 Schema |                 Name                  | Type  |  Owner  
--------+---------------------------------------+-------+---------
 public | averylongtablenamethatseemstoneverend | table | jignesh
 public | customer                              | table | jignesh
 public | ddate                                 | table | jignesh
 public | lineorder                             | table | jignesh
 public | part                                  | table | jignesh
 public | supplier                              | table | jignesh
(6 rows)
```
and 
```
jignesh=# \dt averylongtablenamethatseemstoneverend
                        List of relations
 Schema |                 Name                  | Type  |  Owner  
--------+---------------------------------------+-------+---------
 public | averylongtablenamethatseemstoneverend | table | jignesh
(1 row)
```